### PR TITLE
Add decomp for linalg.cross

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118753
* __->__ #119809

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -1371,6 +1371,29 @@ def addmv(self: Tensor, mat1: Tensor, vec: Tensor, beta: int = 1, alpha: int = 1
...
+    # cross_idx = torch.arange(3, device=a_.device)
+    # return torch.cat([
+    #         a_.index_select(dim, cross_idx[(i + 1) % 3])
+    #         * b_.index_select(dim, cross_idx[(i + 2) % 3])
+    #         - a_.index_select(dim, cross_idx[(i + 2) % 3])
+    #         * b_.index_select(dim, cross_idx[(i + 1) % 3])
+    #         for i in range(3)
+    #     ],
+    #     dim=dim,
+    # )
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:1394]: Remove.

**Review by lezcano — COMMENTED** [review]
This can very well be reference. Mind moving it there + add the relevant OpInfo?

Can you also leave here the benchmarks for reference?

**Comment by amjames** [comment]
There are some unexpected successes to fix, also some accuracy related failures on the gradients to look into. 

I may need to pull the grad tolerance override stuff out of #118753 to use it here.

**Comment by amjames** [comment]
Benchmarks: 
```
[--------- torch.linalg.cross --------]
                   |  Compile  |  Eager
32 threads: ---------------------------
      (8192, 3)    |     13    |    13 
      (3, 8192)    |     13    |    13 
      (16384, 3)   |     13    |    13 
      (3, 16384)   |     13    |    13 
      (32768, 3)   |     13    |    13 
      (3, 32768)   |     13    |    12 
      (65536, 3)   |     13    |    12 
      (3, 65536)   |     13    |    13 
      (131072, 3)  |     18    |    19 
      (3, 131072)  |     21    |    19 
      (262144, 3)  |     33    |    36 
      (3, 262144)  |     54    |    36 
      (524288, 3)  |     65    |    67 
      (3, 524288)  |    107    |    67 

Times are in microseconds (us).
```

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1371,6 +1371,29 @@ def addmv(self: Tensor, mat1: Tensor, vec: Tensor, beta: int = 1, alpha: int = 1
...
+    return torch.cat(
+        [a_1 * b_2 - a_2 * b_1, a_2 * b_0 - a_0 * b_2, a_0 * b_1 - a_1 * b_0], dim=dim
+    )
+    # cross_idx = torch.arange(3, device=a_.device)
+    # return torch.cat([
+    #         a_.index_select(dim, cross_idx[(i + 1) % 3])
+    #         * b_.index_select(dim, cross_idx[(i + 2) % 3])
+    #         - a_.index_select(dim, cross_idx[(i + 2) % 3])
+    #         * b_.index_select(dim, cross_idx[(i + 1) % 3])
+    #         for i in range(3)
```
Comment (peterbell10) [inline torch/_decomp/decompositions.py:1391]: Could we not avoid the cat if you wrote this as `a_.index_select(dim, (cross_idx + 1) % 3)` etc.

Comment (lezcano) [reply torch/_decomp/decompositions.py:1391]: nb. @amjames cat prevents fusion in many cases, so it'd be best to avoid it if we can.

On a separate note, I wonder whether we identify these sort of patterns and we convert them to ModularIndexing... We probably don't, so I don't know if we know how to generate efficient code for that pattern as-is.

Comment (peterbell10) [reply torch/_decomp/decompositions.py:1391]: We don't atm but it could easily be added to `IndexPropagation`

File: torch/_refs/linalg/__init__.py [hunk 3]
```diff
@@ -55,7 +55,19 @@ def _check_norm_dtype(dtype: Optional[torch.dtype], x_dtype: torch.dtype, fn_nam
 
 # Utilities should come BEFORE this import
 from torch._decomp import register_decomposition
+from torch._decomp.decompositions import pw_cast_for_opmath
 
+@register_decomposition(torch._ops.ops.aten.linalg_cross)
+@out_wrapper()
+@pw_cast_for_opmath
```
[the remark this answers is not in the document]
Comment (amjames) [reply torch/_refs/linalg/__init__.py:62]: no,  I will remove it

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (amjames) [reply torch/_decomp/decompositions.py]: In this case `cat` seems to be the only way I can get this to generate a single fused kernel. I have tried a few different ways of creating the result tensor and assigning to slices of it all of them produce multiple kernels. 

The commented out version that I left in here actually generates identical code as the final version I left in. I replaced it because I thought it was a bit easier to read with all other things apparently being equal.

Comment (peterbell10) [reply torch/_decomp/decompositions.py]: @amjames what about the version I suggested above, we don't need to assign any slices.

**Comment by lezcano** [comment]
It looks like you'll need to move `split_with_sizes` to refs as well.

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (peterbell10) [reply torch/_decomp/decompositions.py]: Note that `(cross_idx + 1) % 3` is a tensor so the indexing op will have the same shape as the output.

**Review by lezcano — COMMENTED** [review]
It turns out that IndexPropagation already propagates the `mod` op.
I tried, and

```python
def linalg_cross(a, b, dim: int = -1):
    a, b = torch.broadcast_tensors(a, b)
    idx = torch.arange(3, device=a.device)
    return (a.index_select(dim, (idx + 1) % 3) * b.index_select(dim, (idx + 2) % 3)
          - a.index_select(dim, (idx + 2) % 3) * b.index_select(dim, (idx + 1) % 3))
```
generates great code. @amjames can you confirm?

**Comment by amjames** [comment]
> It turns out that IndexPropagation already propagates the `mod` op. I tried, and
> 
> ```python
> def linalg_cross(a, b, dim: int = -1):
>     a, b = torch.broadcast_tensors(a, b)
>     idx = torch.arange(3, device=a.device)
>     return (a.index_select(dim, (idx + 1) % 3) * b.index_select(dim, (idx + 2) % 3)
>           - a.index_select(dim, (idx + 2) % 3) * b.index_select(dim, (idx + 1) % 3))
> ```
> 
> generates great code. @amjames can you confirm?

Yes, benchmarks are mostly the same, but the a single kernel which is much more compact is generated. Updating now

**Comment by lezcano** [comment]
even more, as discussed above, it avoids using `torch.cat`. `torch.cat` will prevent fusion with other ops after it, while this simpler approach will not.

Also, I would assume this approach does not lag in performance against core even in extreme cases.

**Comment by amjames** [comment]
It appears to sill be falling behind eager 

```
[--------- torch.linalg.cross --------]
                   |  Compile  |  Eager
32 threads: ---------------------------
      (8192, 3)    |     25    |    13 
      (3, 8192)    |     13    |    12 
      (16384, 3)   |     13    |    12 
      (3, 16384)   |     13    |    12 
      (32768, 3)   |     13    |    13 
      (3, 32768)   |     13    |    12 
      (65536, 3)   |     13    |    13 
      (3, 65536)   |     13    |    12 
      (131072, 3)  |     17    |    19 
      (3, 131072)  |     22    |    19 
      (262144, 3)  |     33    |    36 
      (3, 262144)  |     54    |    36 
      (524288, 3)  |     82    |    67 
      (3, 524288)  |    107    |    67 

Times are in microseconds (us).
```

Almost identical to the first benchmark I posted :shrug:

**Comment by lezcano** [comment]
That's fine. As discussed, that last case may be because of the caching hints.

btw, it looks like we can't have this in refs as it uses advanced indexing now. Let's move it back to decomps and test should be passing (try locally and all that).

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 4]
```diff
@@ -2399,6 +2399,12 @@ def make_input():
     #
     # torch.linalg
     #
+    PythonRefInfo(
+        "_refs.linalg.cross",
+        torch_opinfo_name="linalg.cross",
+        supports_out=True,
+        op_db=op_db
+    ),
```
Comment (peterbell10) [inline torch/testing/_internal/opinfo/definitions/linalg.py:2407]: I think just adding a skip is fine here.

Comment (amjames) [reply torch/testing/_internal/opinfo/definitions/linalg.py:2415]: I added the xfails above and they seem to have squashed most of the problems.

The second one does not sit right with me though. The issue is that when the test executes the op [here](https://github.com/pytorch/pytorch/blob/0e6eee3c898c293efc3f172180e9f4d79cc0b13f/test/test_ops.py#L576). The arguments (a, b) are `FakeTensor` instances and the ref implementation creates the `idx` tensor from non-tensor operands. 

The test does not execute with the `FakeTensorMode()` context active, but the failure indicates that it is expected that `idx` will be a `FakeTensor`. Given that a factory is used and there are no `Tensor` arguments, how is that supposed to be possible? Is there a solution I have missed or is the xfail here correct?

Comment (lezcano) [reply torch/testing/_internal/opinfo/definitions/linalg.py:2415]: Ag, yeah, I guess we never hit this issue because we don't do any advanced indexing in refs. I guess the xfail + a comment is fine. Just make sure the error checks match those from eager.

**Review by lezcano — APPROVED** [review]

File: test/inductor/test_torchinductor_opinfo.py [hunk 5]
```diff
@@ -315,6 +313,8 @@ def wrapper_noop_set_seed(op, *args, **kwargs):
     "empty_strided": {"assert_equal": False},
     "new_empty_strided": {"assert_equal": False},
     "randn": {"assert_equal": False},
+    ("cross", "cuda", f16): {"grad_atol": 5e-4, "grad_rtol": 0.003},
+    ("linalg.cross", "cuda", f16): {"grad_atol": 5e-4, "grad_rtol": 0.003},
```
Comment (peterbell10) [inline test/inductor/test_torchinductor_opinfo.py:317]: NIT: Grad tolerance is certainly better than a skip, but I wonder if `reference_in_float=True` would also solve these issues?

**Review by peterbell10 — APPROVED** [review]

File: torch/_refs/linalg/__init__.py [hunk 6]
```diff
@@ -57,6 +66,25 @@ def _check_norm_dtype(dtype: Optional[torch.dtype], x_dtype: torch.dtype, fn_nam
 from torch._decomp import register_decomposition
 
 
+@register_decomposition(torch._ops.ops.aten.linalg_cross)
+@out_wrapper()
+def cross(a: Tensor, b: Tensor, dim: int = -1):
```
Comment (peterbell10) [inline torch/_refs/linalg/__init__.py:71]: Given the CI failures I say we should just use opmath. There's no reason we shouldn't want higher precision than eager.
